Fix limits on `scale_colour_gradient2()`

foo-bar-baz-qux opened the pull request:
Fixes #2230. 

Limits are now imposed before rescaling so points outside the limits don't reappear.

hadley commented:
Could you add a little test for this please?

foo-bar-baz-qux commented:
Test added, and rebased to the master branch.

hadley commented on tests/testthat/test-scale-gradient.R:
```diff
@@ -0,0 +1,18 @@
+                                mid = "gray80",
+                                midpoint = 3, limits = c(2, 6),
+                                na.value = "orange")
+  p_data <- ggplot_build(p)$data[[1]]
```
Can you please use `layer_data()` here

hadley replied:
And I think a made up data set would make the test simpler. Maybe something like this?

```R
library(ggplot2)
df <- data.frame(x = c(0, 1, 2))

p <- ggplot(df, aes(x, 1, fill = x)) +
  geom_col() + 
  scale_fill_gradient2(limits = c(-1, 1), na.value = "orange")

layer_data(p)$fill
```

hadley commented:
Are you interested in finishing this off? No problems if you don't have time; just let me know I'll get it over the finish line.

foo-bar-baz-qux commented:
Yep, I can make those fixes.

foo-bar-baz-qux commented:
Updated with a simpler test, thanks for the example. I've modified the `midpoint` in the example so it fails on current master build but passes on this version.

hadley commented:
Thanks!
